### DISCUSSION TITLE
[Chapter 06] 종합 질문

### DISCUSSION THREAD

**Discussion by ahngilwoong** [cowaine/spring-boot-for-msa#53] [Chapter 06. Building a Web Application Server] [answered]
1. Table명은 단수형이 적절할까요? 복수형이 적절할까요? 
2. Spring 환경에서의 다국어 처리 방법에 대해서 잘 설명이 되어있는데 완벽한 다국어 처리를 위해선 결국 DB도 다국어도 여러개 둬야하는건가요?
3. 국제화를 경험해보신 분이 있다면 Locale를 처리하는 방법은 Accept-Language 헤더, QueryParam????등 어떤 방식을 쓰는게 이상적일까요

**Comment by BvrPark** [comment, marked as answer]
### A1. 
제가 알기론 영어에서 어떤 대상이 단수일 가능성도, 복수일 가능성도 존재할 때, 뒤에 s를 붙인다고 알고있습니다.
그리고 코딩 용어들 자체가 다 영어이니 Table명은 복수가 맞지 않을까?라고 생각이 듭니다.

### A2. 
이건 저도 궁금해서 찾아봤습니다.
보통은 다국어를 DB까지는 적용을 하지 않는다고 합니다.
(DB에서 가져온 데이터는 항상 동일한 입력 조건에는 동일한 결과가 나와야 되는 특성 때문이라고 합니다.)

그래서 보통 어플리케이션에서 다국어 팩 파일 정도만 사용하면 된다고 합니다.(properties, xml, json 등)
하지만 상황에 따라서 DB에서도 다국어 지원을 해야되는 경우에는 크게 4가지 방법이 있다고 합니다.

    1. 각 필드별 칼럼으로 구분 
    2. 다국어가 있으면 row를 새로생성
    3. 여러 항목을 가진 다국어 테이블을 하나만 만들기
    4. 다국어 테이블을 새로 생성
    
그리고 이 중에서도 보통 4번 방법을 쓰는 것 같습니다.
4번 방법의 진행과정은 이렇습니다.
    
    1. 테이블 명을 key로 해서 그에 따른 지원 언어(locale)을 저장.
    2. 사이트에 접속했을 때, locale을 스캔해서 각 테이블 별로 읽어온 locale에 존재유무를 판단.
    3. 존재하면 해당 locale을 쓰고 없으면 default locale(영어, 한국어 등)을 설정하여 default locale을 사용
    
대신 이 방법은 pk가 바뀝니다.
쇼핑몰을 예로 들면, 동일한 게시물이어도 locale 별로 다르게 나와야 되기 때문에 
게시물 pk와 locale을 묶어서 복합키로 pk를 설정해야 될 것 같습니다. 

### A3.
이건 저도 안써봐서 제가 질문을 제대로 이해하고 있는지는 잘 모르겠지만 
인터셉터를 이용해서 웹 요청에 따라서 바로 Locale을 처리할 수 있지 않을까요...?
저도 궁금합니다!
    
### 참고사이트
- [stackoverflow](https://stackoverflow.com/questions/929410/what-are-best-practices-for-multi-language-database-design)
- [어느 외국인의 블로그](https://medium.com/walkin/database-internationalization-i18n-localization-l10n-design-patterns-94ff372375c6)
